File: backend/app/api/knowledge_api.py
"""知识库 HTTP 路由：元数据 CRUD、文档列表与上传。"""

from typing import Annotated

from fastapi import APIRouter, Depends, File, HTTPException, Query, Response, UploadFile, status
from fastapi.responses import JSONResponse

from app.core.deps import DBManagerDep, MinioObjectStoreDep
from app.schemas.knowledge import (
    KnowledgeBaseOut,
    KnowledgeCreateBody,
    KnowledgeDocumentChunksData,
    KnowledgeDocumentIngestAccepted,
    KnowledgeDocumentListData,
    KnowledgeDocumentOut,
    KnowledgeDocumentPatchBody,
    KnowledgeListData,
    KnowledgeSearchBody,
    KnowledgeSearchData,
    KnowledgeUpdateBody,
    KnowledgeVectorBuildData,
)
from app.schemas.response import ApiResponse
from app.services.knowledge_svc import (
    ChunkRepositoryDep,
    KnowledgeBaseService,
    KnowledgeDocumentService,
)


def get_knowledge_base_service(db_manager: DBManagerDep) -> KnowledgeBaseService:
    return KnowledgeBaseService(db_manager)


def get_knowledge_document_service(
    db_manager: DBManagerDep,
    minio_store: MinioObjectStoreDep,
    chunk_repository: ChunkRepositoryDep,
) -> KnowledgeDocumentService:
    return KnowledgeDocumentService(db_manager, minio_store, chunk_repository)


KnowledgeBaseServiceDep = Annotated[KnowledgeBaseService, Depends(get_knowledge_base_service)]
KnowledgeDocumentServiceDep = Annotated[
    KnowledgeDocumentService,
    Depends(get_knowledge_document_service),
]

router = APIRouter()


@router.get(
    "",
    response_model=ApiResponse[KnowledgeListData],
    summary="知识库列表（分页）",
    operation_id="get_knowledge_bases_list",
)
async def list_knowledge_bases(
    svc: KnowledgeBaseServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    q: Annotated[str | None, Query(description="名称或 slug 模糊搜索")] = None,
    status: Annotated[str | None, Query(description="生命周期状态筛选")] = None,
) -> ApiResponse[KnowledgeListData]:
    data = await svc.list_knowledge_bases(page=page, page_size=page_size, q=q, status=status)
    return ApiResponse(message="ok", data=data)


@router.post(
    "",
    response_model=ApiResponse[KnowledgeBaseOut],
    summary="创建知识库",
    operation_id="post_knowledge_bases_create",
)
async def create_knowledge_base_route(
    body: KnowledgeCreateBody,
    svc: KnowledgeBaseServiceDep,
) -> ApiResponse[KnowledgeBaseOut]:
    try:
        out = await svc.create_knowledge_base(body)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return ApiResponse(message="ok", data=out)


@router.post(
    "/{kb_id}/vectors/build",
    response_model=ApiResponse[KnowledgeVectorBuildData],
    summary="从 Mongo 已有分片构建向量并写入 Milvus（不重跑 ingest；需 Milvus 与 vector/hybrid + embedding）",
    operation_id="post_knowledge_vectors_build",
)
async def build_knowledge_vectors_route(
    kb_id: int,
    svc: KnowledgeDocumentServiceDep,
) -> ApiResponse[KnowledgeVectorBuildData]:
    try:
        data = await svc.build_vectors_from_mongo_chunks(kb_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.post(
    "/{kb_id}/search",
    response_model=ApiResponse[KnowledgeSearchData],
    summary="知识库检索（可选 retrieval_override / recall_limit / rrf_k；向量/混合在 Milvus 未接入时降级为关键词）",
    operation_id="post_knowledge_kb_search",
)
async def search_knowledge_route(
    kb_id: int,
    body: KnowledgeSearchBody,
    svc: KnowledgeDocumentServiceDep,
) -> ApiResponse[KnowledgeSearchData]:
    try:
        data = await svc.search_in_kb(
            kb_id,
            q=body.q,
            limit=body.limit,
            doc_id=body.doc_id,
            retrieval_override=body.retrieval_override,
            recall_limit=body.recall_limit,
            rrf_k=body.rrf_k,
        )
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.get(
    "/{kb_id}/documents",
    response_model=ApiResponse[KnowledgeDocumentListData],
    summary="知识库文档列表（分页）",
    operation_id="get_knowledge_documents_list",
)
async def list_knowledge_documents(
    kb_id: int,
    svc: KnowledgeDocumentServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    q: Annotated[str | None, Query(description="文件名模糊搜索")] = None,
    chunk_left_panel: Annotated[
        bool,
        Query(
            description=(
                "为 true 时返回切块步骤左侧「待配置」列表：排除「已 indexed 且已设文档级 chunk_method」；"
                "恢复默认后仅 indexed 且无覆盖的文档会重新出现在此列表"
            ),
        ),
    ] = False,
) -> ApiResponse[KnowledgeDocumentListData]:
    try:
        data = await svc.list_documents(
            kb_id,
            page=page,
            page_size=page_size,
            q=q,
            chunk_left_panel=chunk_left_panel,
        )
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.post(
    "/{kb_id}/documents/upload",
    response_model=ApiResponse[KnowledgeDocumentIngestAccepted],
    summary="上传文档（本地落盘；切块/索引由后续任务处理）",
    operation_id="post_knowledge_document_upload",
    responses={
        202: {
            "description": "已接受异步索引任务（与 200 相同响应体，便于客户端按状态码区分）",
            "model": ApiResponse[KnowledgeDocumentIngestAccepted],
        },
    },
)
async def upload_knowledge_document_route(
    kb_id: int,
    svc: KnowledgeDocumentServiceDep,
    file: Annotated[UploadFile, File(description="待上传文件")],
    async_: Annotated[
        bool,
        Query(
            alias="async",
            description="为 true 时返回 HTTP 202（仍返回 document + task_id）",
        ),
    ] = False,
) -> ApiResponse[KnowledgeDocumentIngestAccepted] | JSONResponse:
    raw = await file.read()
    try:
        accepted = await svc.upload_document(
            kb_id,
            original_filename=file.filename or "",
            content_type=file.content_type,
            data=raw,
        )
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    except RuntimeError as e:
        raise HTTPException(status_code=500, detail=str(e)) from e
    body = ApiResponse(message="accepted" if async_ else "ok", data=accepted)
    if async_:
        return JSONResponse(status_code=202, content=body.model_dump(mode="json"))
    return body


@router.patch(
    "/{kb_id}/documents/{doc_id}",
    response_model=ApiResponse[KnowledgeDocumentOut],
    summary="更新文档（文档级切块策略覆盖；chunk_method 置 null 表示恢复知识库默认）",
    operation_id="patch_knowledge_document_by_id",
)
async def patch_knowledge_document_route(
    kb_id: int,
    doc_id: int,
    body: KnowledgeDocumentPatchBody,
    svc: KnowledgeDocumentServiceDep,
) -> ApiResponse[KnowledgeDocumentOut]:
    try:
        data = await svc.patch_document(kb_id, doc_id, body)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.post(
    "/{kb_id}/documents/{doc_id}/ingest",
    response_model=ApiResponse[KnowledgeDocumentIngestAccepted],
    summary="重新触发文档 ingest（删除同版本幂等任务后再次入队）",
    operation_id="post_knowledge_document_reingest",
    responses={
        202: {
            "description": "已接受异步索引任务",
            "model": ApiResponse[KnowledgeDocumentIngestAccepted],
        },
    },
)
async def requeue_knowledge_document_ingest_route(
    kb_id: int,
    doc_id: int,
    svc: KnowledgeDocumentServiceDep,
    async_: Annotated[
        bool,
        Query(
            alias="async",
            description="为 true 时返回 HTTP 202",
        ),
    ] = False,
) -> ApiResponse[KnowledgeDocumentIngestAccepted] | JSONResponse:
    try:
        accepted = await svc.requeue_document_ingest(kb_id, doc_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    body = ApiResponse(message="accepted" if async_ else "ok", data=accepted)
    if async_:
        return JSONResponse(status_code=202, content=body.model_dump(mode="json"))
    return body


@router.delete(
    "/{kb_id}/documents/{doc_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除文档（软删；清理 Mongo 分片并尽力删除 MinIO 对象）",
    operation_id="delete_knowledge_document_by_id",
)
async def delete_knowledge_document_route(
    kb_id: int,
    doc_id: int,
    svc: KnowledgeDocumentServiceDep,
) -> Response:
    try:
        await svc.delete_document(kb_id, doc_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get(
    "/{kb_id}/documents/{doc_id}/chunks",
    response_model=ApiResponse[KnowledgeDocumentChunksData],
    summary="文档分片列表（Mongo，按 content_version 分页）",
    operation_id="get_knowledge_document_chunks",
)
async def list_document_chunks_route(
    kb_id: int,
    doc_id: int,
    svc: KnowledgeDocumentServiceDep,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
    content_version: Annotated[
        str | None,
        Query(description="留空则使用文档当前 content_version"),
    ] = None,
) -> ApiResponse[KnowledgeDocumentChunksData]:
    try:
        data = await svc.list_document_chunks(
            kb_id,
            doc_id,
            content_version=content_version,
            page=page,
            page_size=page_size,
        )
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.get(
    "/{kb_id}",
    response_model=ApiResponse[KnowledgeBaseOut],
    summary="知识库详情",
    operation_id="get_knowledge_base_by_id",
)
async def get_knowledge_base_route(
    kb_id: int,
    svc: KnowledgeBaseServiceDep,
) -> ApiResponse[KnowledgeBaseOut]:
    try:
        data = await svc.get_knowledge_base(kb_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.patch(
    "/{kb_id}",
    response_model=ApiResponse[KnowledgeBaseOut],
    summary="更新知识库元数据与切块策略",
    operation_id="patch_knowledge_base_by_id",
)
async def patch_knowledge_base_route(
    kb_id: int,
    body: KnowledgeUpdateBody,
    svc: KnowledgeBaseServiceDep,
) -> ApiResponse[KnowledgeBaseOut]:
    try:
        data = await svc.update_knowledge_base(kb_id, body)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return ApiResponse(message="ok", data=data)


@router.delete(
    "/{kb_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除知识库（软删；派生存储清理见异步任务）",
    operation_id="delete_knowledge_base_by_id",
)
async def delete_knowledge_base_route(
    kb_id: int,
    svc: KnowledgeBaseServiceDep,
) -> Response:
    try:
        await svc.delete_knowledge_base(kb_id)
    except LookupError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/knowledgeApi.ts
/**
 * 知识库 API（GET/POST /api/knowledge…），信封 message + data。
 * 路由实现见后端 ``app.api.knowledge_api``，编排见 ``app.services.knowledge_svc``。
 */

import { parseApiEnvelope } from './parseEnvelope'

export type KnowledgeBaseOut = {
  id: number
  name: string
  slug: string
  description: string | null
  storage_type: string
  retrieval_type: string
  status: string
  embedding_model_config_id: number | null
  milvus_collection: string | null
  minio_prefix: string | null
  chunk_method: string
  chunk_size: number
  chunk_overlap: number
  chunk_separator: string | null
  config_json: Record<string, unknown> | null
  /** 结构化切块策略（与 ``config_json.chunk_strategy`` 及列字段一致） */
  chunk_strategy: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type KnowledgeListData = {
  items: KnowledgeBaseOut[]
  total: number
  page: number
  page_size: number
}

export type KnowledgeDocumentOut = {
  id: number
  kb_id: number
  filename: string
  object_key: string | null
  size_bytes: number | null
  mime: string | null
  sha256: string | null
  status: string
  content_version: string
  chunk_count: number
  chunk_method: string | null
  chunk_size: number | null
  chunk_overlap: number | null
  chunk_separator: string | null
  created_at: string
  updated_at: string
}

/** POST 上传 / 重新 ingest 接受体，与 OpenAPI ``KnowledgeDocumentIngestAccepted`` 一致 */
export type KnowledgeDocumentIngestAccepted = {
  document: KnowledgeDocumentOut
  /** 幂等去重未新建任务时为 null */
  task_id: string | null
}

/** PATCH /api/knowledge/{kb_id}/documents/{doc_id}；chunk_method 显式 null 表示恢复知识库默认 */
export type KnowledgeDocumentPatchBody = {
  chunk_method?: string | null
  chunk_size?: number | null
  chunk_overlap?: number | null
  chunk_separator?: string | null
}

export type KnowledgeDocumentListData = {
  items: KnowledgeDocumentOut[]
  total: number
  page: number
  page_size: number
}

/** POST /api/knowledge/{kb_id}/search */
export type KnowledgeSearchBody = {
  q: string
  limit?: number
  /** 仅在该文档的分片中检索（与全库检索共用同一接口） */
  doc_id?: number | null
  /** 覆盖知识库 retrieval_type；不传则与库配置一致 */
  retrieval_override?: 'keyword' | 'vector' | 'hybrid' | null
  /** 各子路召回上限（≥ limit，≤100）；混合/向量扩大预取；不传则服务端默认 */
  recall_limit?: number | null
  /** 混合检索 RRF 常数 k（1～200）；不传则服务端默认 60 */
  rrf_k?: number | null
}

export type KnowledgeSearchHit = {
  doc_id: number
  chunk_index: number | null
  text_snippet: string | null
  filename: string | null
  match_type: 'chunk' | 'filename' | 'keyword' | 'vector'
  /** 混合路为 RRF 合分；关键词/向量单路多为空 */
  score?: number | null
}

export type KnowledgeSearchData = {
  items: KnowledgeSearchHit[]
  total: number
  /** 知识库索引配置中的检索类型 */
  configured_retrieval: 'keyword' | 'vector' | 'hybrid'
  /** 本次请求实际执行的检索类型 */
  applied_retrieval: 'keyword' | 'vector' | 'hybrid'
  /** 降级或说明 */
  note?: string | null
}

/** POST /api/knowledge/{kb_id}/vectors/build — 从 Mongo 已有分片构建 Milvus 向量（不重跑 ingest） */
export type KnowledgeVectorBuildData = {
  built: number
  skipped: number
  errors: string[]
}

export type KnowledgeChunkOut = {
  chunk_index: number
  text: string
  created_at: string | null
}

export type KnowledgeDocumentChunksData = {
  items: KnowledgeChunkOut[]
  total: number
  page: number
  page_size: number
  doc_id: number
  filename: string | null
  content_version: string
  document_status: string | null
  /** total=0 时后端给出的说明 */
  empty_hint?: string | null
}

export type KnowledgeCreateBody = {
  name: string
  slug?: string | null
  description?: string | null
}

/** PATCH /api/knowledge/{kb_id}；字段与 OpenAPI 对齐，均为可选 */
export type KnowledgeUpdateBody = {
  name?: string | null
  description?: string | null
  storage_type?: 'keyword' | 'vector' | 'hybrid'
  retrieval_type?: 'keyword' | 'vector' | 'hybrid'
  status?: 'empty' | 'draft' | 'ready' | 'processing' | 'failed'
  embedding_model_config_id?: number | null
  milvus_collection?: string | null
  minio_prefix?: string | null
  chunk_method?: string | null
  chunk_size?: number | null
  chunk_overlap?: number | null
  chunk_separator?: string | null
  config_json?: Record<string, unknown> | null
  /** PATCH 时与顶栏 chunk_* 同步写入 ``config_json.chunk_strategy`` */
  chunk_strategy?: Record<string, unknown> | null
}

const jsonHeaders = { 'Content-Type': 'application/json', Accept: 'application/json' }

const base = '/api/knowledge'

export async function fetchKnowledgeBasesList(params: {
  page?: number
  page_size?: number
  q?: string
  status?: string
}): Promise<KnowledgeListData> {
  const sp = new URLSearchParams()
  if (params.page != null) sp.set('page', String(params.page))
  if (params.page_size != null) sp.set('page_size', String(params.page_size))
  if (params.q?.trim()) sp.set('q', params.q.trim())
  if (params.status?.trim()) sp.set('status', params.status.trim())
  const q = sp.toString()
  const url = q ? `${base}?${q}` : base
  const res = await fetch(url, { headers: { Accept: 'application/json' } })
  return parseApiEnvelope<KnowledgeListData>(res)
}

export async function fetchKnowledgeBase(kbId: number): Promise<KnowledgeBaseOut> {
  const res = await fetch(`${base}/${kbId}`, { headers: { Accept: 'application/json' } })
  return parseApiEnvelope<KnowledgeBaseOut>(res)
}

export async function postBuildKnowledgeVectors(kbId: number): Promise<KnowledgeVectorBuildData> {
  const res = await fetch(`${base}/${kbId}/vectors/build`, {
    method: 'POST',
    headers: { Accept: 'application/json' },
  })
  return parseApiEnvelope<KnowledgeVectorBuildData>(res)
}

export async function updateKnowledgeBase(
  kbId: number,
  body: KnowledgeUpdateBody,
): Promise<KnowledgeBaseOut> {
  const res = await fetch(`${base}/${kbId}`, {
    method: 'PATCH',
    headers: jsonHeaders,
    body: JSON.stringify(body),
  })
  return parseApiEnvelope<KnowledgeBaseOut>(res)
}

export async function createKnowledgeBase(body: KnowledgeCreateBody): Promise<KnowledgeBaseOut> {
  const res = await fetch(base, {
    method: 'POST',
    headers: jsonHeaders,
    body: JSON.stringify(body),
  })
  return parseApiEnvelope<KnowledgeBaseOut>(res)
}

export async function deleteKnowledgeBase(kbId: number): Promise<void> {
  const res = await fetch(`${base}/${kbId}`, { method: 'DELETE', headers: { Accept: 'application/json' } })
  if (res.status === 204) return
  await parseApiEnvelope<unknown>(res)
}

export async function fetchKnowledgeDocuments(params: {
  kbId: number
  page?: number
  page_size?: number
  q?: string
  /**
   * 为 true 时切块第 2 步左侧「待配置」列表：排除「已 indexed 且已有文档级切块覆盖」；
   * 恢复默认后重新出现在左侧。
   */
  chunk_left_panel?: boolean
}): Promise<KnowledgeDocumentListData> {
  const sp = new URLSearchParams()
  if (params.page != null) sp.set('page', String(params.page))
  if (params.page_size != null) sp.set('page_size', String(params.page_size))
  if (params.q?.trim()) sp.set('q', params.q.trim())
  if (params.chunk_left_panel) sp.set('chunk_left_panel', 'true')
  const q = sp.toString()
  const url = q ? `${base}/${params.kbId}/documents?${q}` : `${base}/${params.kbId}/documents`
  const res = await fetch(url, { headers: { Accept: 'application/json' } })
  return parseApiEnvelope<KnowledgeDocumentListData>(res)
}

export async function uploadKnowledgeDocument(
  kbId: number,
  file: File,
): Promise<KnowledgeDocumentIngestAccepted> {
  const fd = new FormData()
  fd.append('file', file)
  const res = await fetch(`${base}/${kbId}/documents/upload`, {
    method: 'POST',
    body: fd,
    headers: { Accept: 'application/json' },
  })
  return parseApiEnvelope<KnowledgeDocumentIngestAccepted>(res)
}

export async function deleteKnowledgeDocument(kbId: number, docId: number): Promise<void> {
  const res = await fetch(`${base}/${kbId}/documents/${docId}`, {
    method: 'DELETE',
    headers: { Accept: 'application/json' },
  })
  if (res.status === 204) return
  await parseApiEnvelope<unknown>(res)
}

export async function patchKnowledgeDocument(
  kbId: number,
  docId: number,
  body: KnowledgeDocumentPatchBody,
): Promise<KnowledgeDocumentOut> {
  const res = await fetch(`${base}/${kbId}/documents/${docId}`, {
    method: 'PATCH',
    headers: jsonHeaders,
    body: JSON.stringify(body),
  })
  return parseApiEnvelope<KnowledgeDocumentOut>(res)
}

/** POST /api/knowledge/{kb_id}/documents/{doc_id}/ingest — 重新触发 ingest */
export async function postRequeueDocumentIngest(
  kbId: number,
  docId: number,
): Promise<KnowledgeDocumentIngestAccepted> {
  const res = await fetch(`${base}/${kbId}/documents/${docId}/ingest`, {
    method: 'POST',
    headers: { Accept: 'application/json' },
  })
  return parseApiEnvelope<KnowledgeDocumentIngestAccepted>(res)
}

export async function searchKnowledgeInKb(
  kbId: number,
  body: KnowledgeSearchBody,
): Promise<KnowledgeSearchData> {
  const payload: Record<string, unknown> = {
    q: body.q.trim(),
    limit: body.limit ?? 20,
  }
  if (body.doc_id != null && Number.isFinite(Number(body.doc_id))) {
    payload.doc_id = Number(body.doc_id)
  }
  if (body.retrieval_override != null) {
    payload.retrieval_override = body.retrieval_override
  }
  if (body.recall_limit != null && Number.isFinite(Number(body.recall_limit))) {
    payload.recall_limit = Number(body.recall_limit)
  }
  if (body.rrf_k != null && Number.isFinite(Number(body.rrf_k))) {
    payload.rrf_k = Number(body.rrf_k)
  }
  const res = await fetch(`${base}/${kbId}/search`, {
    method: 'POST',
    headers: jsonHeaders,
    body: JSON.stringify(payload),
  })
  return parseApiEnvelope<KnowledgeSearchData>(res)
}

export async function fetchDocumentChunks(params: {
  kbId: number
  docId: number
  page?: number
  page_size?: number
  content_version?: string | null
}): Promise<KnowledgeDocumentChunksData> {
  const sp = new URLSearchParams()
  if (params.page != null) sp.set('page', String(params.page))
  if (params.page_size != null) sp.set('page_size', String(params.page_size))
  if (params.content_version?.trim()) sp.set('content_version', params.content_version.trim())
  const q = sp.toString()
  const url = q
    ? `${base}/${params.kbId}/documents/${params.docId}/chunks?${q}`
    : `${base}/${params.kbId}/documents/${params.docId}/chunks`
  const res = await fetch(url, { headers: { Accept: 'application/json' } })
  return parseApiEnvelope<KnowledgeDocumentChunksData>(res)
}
